Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean

```diff
@@ -0,0 +1,201 @@
+    ZMod.toAddCircle.comp <| .mulLeft x
+
+@[simp] lemma zmod_intCast (x y : ℤ) : zmod n x y = exp (2 * π * (x * y / n)) := by
+  simp [zmod, ← Int.cast_mul x y, -Int.cast_mul, ZMod.toAddCircle_intCast,
+    AddCircle.toCircle_apply_mk]
+
+@[simp] lemma zmod_zero : zmod n 0 = 1 :=
+  DFunLike.ext _ _ <| by simp [ZMod.intCast_surjective.forall, zmod]
+
+variable {n}
+
+@[simp] lemma zmod_add : ∀ x y : ZMod n, zmod n (x + y) = zmod n x * zmod n y := by
+  simp [DFunLike.ext_iff, ← Int.cast_add, zmod, add_mul, add_div, map_add_eq_mul]
+
+lemma zmod_injective : Injective (zmod n) := by
+  simp_rw [Injective, ZMod.intCast_surjective.forall]
+  rintro x y h
+  replace hn : (n : ℝ) ≠ 0 := by have := NeZero.ne n; positivity
+  simpa [pi_ne_zero, exp_inj, hn, CharP.intCast_eq_intCast (ZMod n) n] using
+    (zmod_intCast ..).symm.trans <| (DFunLike.congr_fun h ((1 : ℤ) : ZMod n)).trans <|
+      zmod_intCast ..
+
+@[simp] lemma zmod_inj {x y : ZMod n} : zmod n x = zmod n y ↔ x = y := zmod_injective.eq_iff
+
+/-- `AddChar.zmod` bundled as an `AddChar`. -/
+def zmodHom : AddChar (ZMod n) (AddChar (ZMod n) Circle) where
+  toFun := zmod n
+  map_zero_eq_one' := by simp
+  map_add_eq_mul' := by simp
+
+/-- Character on a product of `ZMod`s given by `x ↦ ∏ i, e ^ (2 * π * I * x i * y / n)`. -/
+private def mkZModAux {ι : Type} [DecidableEq ι] (n : ι → ℕ) [∀ i, NeZero (n i)]
+    (u : ∀ i, ZMod (n i)) : AddChar (⨁ i, ZMod (n i)) Circle :=
+  AddChar.directSum fun i ↦ zmod (n i) (u i)
+
+private lemma mkZModAux_injective {ι : Type} [DecidableEq ι] {n : ι → ℕ} [∀ i, NeZero (n i)] :
+    Injective (mkZModAux n) :=
+  AddChar.directSum_injective.comp fun f g h ↦ by simpa [funext_iff] using h
+
+/-- The circle-valued characters of a finite abelian group are the same as its complex-valued
+characters. -/
+def circleEquivComplex [Finite α] : AddChar α Circle ≃+ AddChar α ℂ where
+  toFun ψ := toMonoidHomEquiv.symm <| coeHom.comp ψ.toMonoidHom
+  invFun ψ :=
+    { toFun := fun a ↦ (⟨ψ a, mem_sphere_zero_iff_norm.2 <| ψ.norm_apply _⟩ : Circle)
+      map_zero_eq_one' := by simp [Circle]
+      map_add_eq_mul' := fun a b ↦ by ext : 1; simp [map_add_eq_mul] }
+  left_inv ψ := by ext : 1; simp
+  right_inv ψ := by ext : 1; simp
+  map_add' ψ χ := rfl
+
+@[simp] lemma card_eq [Fintype α] : card (AddChar α ℂ) = card α := by
+  obtain ⟨ι, _, n, hn, ⟨e⟩⟩ := AddCommGroup.equiv_directSum_zmod_of_finite' α
+  classical
+  have hn' i : NeZero (n i) := by have := hn i; exact ⟨by positivity⟩
+  let f : α → AddChar α ℂ := fun a ↦ coeHom.compAddChar ((mkZModAux n <| e a).compAddMonoidHom e)
+  have hf : Injective f := circleEquivComplex.injective.comp
+    ((compAddMonoidHom_injective_left _ e.surjective).comp <| mkZModAux_injective.comp <|
+      DFunLike.coe_injective.comp <| e.injective.comp Additive.ofMul.injective)
+  exact (card_addChar_le _ _).antisymm (Fintype.card_le_of_injective _ hf)
```

Comment:
With the additional imports
```
import Mathlib.GroupTheory.FiniteAbelian.Duality
import Mathlib.RingTheory.RootsOfUnity.AlgebraicallyClosed
import Mathlib.Analysis.Complex.Polynomial.Basic
```
one can use the results on homomorphisms from finite abelian multiplicative groups into monoids with enough roots of unity, as follows.
```
variable (α) in
/-- The additive isomorphism from `AddChar α M` to homomorphisms from `Multiplicative α` to `M`. -/
def mulEquivMultiplicativeMonoidHom (M) [CommMonoid M] :
    Multiplicative (AddChar α M) ≃* (Multiplicative α →* M) :=
  let e : Multiplicative (AddChar α M) ≃* AddChar α M :=
    MulEquiv.mk' Additive.ofMul fun _ _ ↦ rfl
  e.trans <| AddChar.toMonoidHomMulEquiv

variable (α) in
/-- The group of complex-valued additive characters on a finite abelian group is
(noncanonically) isomorphic to itself. -/
lemma addEquiv_self [Finite α] : Nonempty (AddChar α ℂ ≃+ α) :=
  let e := (CommGroup.mulEquiv_monoidHom_of_hasEnoughRootsOfUnity (Multiplicative α) ℂ).some
  let e₀ : (Multiplicative α →* ℂ) ≃* (Multiplicative α →* ℂˣ) := {
    toFun := MonoidHom.toHomUnits
    invFun := fun φ ↦ (Units.coeHom _).comp φ
    left_inv := fun φ ↦ by ext; simp
    right_inv := fun φ ↦ by ext; simp
    map_mul' := fun φ ψ ↦ by ext; simp
  }
  let e' := (mulEquivMultiplicativeMonoidHom α ℂ).trans <| e₀.trans e.symm
  ⟨AddEquiv.toMultiplicative.symm e'⟩

@[simp] lemma card_eq [Fintype α] : card (AddChar α ℂ) = card α :=
  Fintype.card_congr (addEquiv_self α).some.toEquiv
```

Reply:
Do you think there should be a TODO in this file so it doesn't get forgotten? Or do you think the linkage here isn't worth it in general

Reply:
Left a TODO